### PR TITLE
NEW module user rights enhancement

### PR DESCRIPTION
**Pull request by hregis**

### CODE REVIEW

**Comment by hregis** [comment]
@eldy it's ok for you for the first changes ?

File: htdocs/install/mysql/migration/19.0.0-20.0.0.sql [hunk 1]
```diff
@@ -290,3 +290,6 @@ ALTER TABLE llx_receptiondet_batch CHANGE COLUMN fk_commande fk_element integer;
 ALTER TABLE llx_receptiondet_batch CHANGE COLUMN fk_commandefourndet fk_elementdet integer;
 
 ALTER TABLE llx_supplier_proposaldet MODIFY ref_fourn VARCHAR(128) NULL;
+
+ALTER TABLE llx_rights_def ADD COLUMN module_origin varchar(64) AFTER module;
```
Comment (eldy) [inline htdocs/install/mysql/migration/19.0.0-20.0.0.sql:294]: Difference between module and module_origin is not clear.
One is the name of module the permission is for
The other is the module that provide the permission. But which one is which one ?

**Review by eldy — COMMENTED** [review]
Difference between module and module_origin is not clear.
One is the name of module the permission is for
The other is the module that provide the permission. But which one is which one ?

File: htdocs/core/modules/DolibarrModules.class.php [hunk 2]
```diff
@@ -1871,46 +1879,95 @@ public function insert_permissions($reinitadminperms = 0, $force_entity = null,
 			if ($obj !== null && !empty($obj->value) && !empty($this->rights)) {
 				include_once DOL_DOCUMENT_ROOT.'/user/class/user.class.php';
 
+				// TODO right parameters with an integer is deprecated
```
Comment (mdeweerd) [inline htdocs/core/modules/DolibarrModules.class.php:1882]: Spelling/wording improvement: '`rights`parameters with integer indexes are deprecated'

File: htdocs/core/modules/DolibarrModules.class.php [hunk 3]
```diff
@@ -1871,46 +1879,95 @@ public function insert_permissions($reinitadminperms = 0, $force_entity = null,
...
 				// If the module is active
 				foreach ($this->rights as $key => $value) {
-					$r_id       = $this->rights[$key][0];	// permission id in llx_rights_def (not unique because primary key is couple id-entity)
-					$r_desc     = $this->rights[$key][1];
-					$r_type     = isset($this->rights[$key][2]) ? $this->rights[$key][2] : '';
-					$r_def      = empty($this->rights[$key][3]) ? 0 : $this->rights[$key][3];
-					$r_perms    = $this->rights[$key][4];
-					$r_subperms = isset($this->rights[$key][5]) ? $this->rights[$key][5] : '';
-					$r_modul = empty($this->rights_class) ? strtolower($this->name) : $this->rights_class;
+					$r_id	= (isset($this->rights[$key][Rights::KEY_OLD_ID]) ? $this->rights[$key][Rights::KEY_OLD_ID] : $this->rights[$key][Rights::KEY_ID]);	// permission id in llx_rights_def (not unique because primary key is couple id-entity)
```
Comment (mdeweerd) [inline htdocs/core/modules/DolibarrModules.class.php:1896]: Maybe simplify with ?? operator?

File: htdocs/core/modules/DolibarrModules.class.php [hunk 4]
```diff
@@ -1871,46 +1879,95 @@ public function insert_permissions($reinitadminperms = 0, $force_entity = null,
...
+
+					if (isset($this->rights[$key][Rights::KEY_MODULE])) {
+						// name of the module to which the right must be applied
+						$r_module = $this->rights[$key][Rights::KEY_MODULE];
+						// name of the module from which the right comes
+						$r_module_origin = (empty($this->rights_class) ? strtolower($this->name) : $this->rights_class);
+					}
+
+					// entity for this right ((default: current entity id)
+					$r_entity = ((isset($this->rights[$key][Rights::KEY_ENTITY]) && ($this->rights[$key][Rights::KEY_ENTITY] == 0 || $this->rights[$key][Rights::KEY_ENTITY] == 'all' )) ? 0 : $entity);
```
Comment (mdeweerd) [inline htdocs/core/modules/DolibarrModules.class.php:1927]: I would even have a constant for 'all'.

File: htdocs/core/modules/DolibarrModules.class.php [hunk 5]
```diff
@@ -1871,46 +1879,95 @@ public function insert_permissions($reinitadminperms = 0, $force_entity = null,
...
+					$r_id	= (isset($this->rights[$key][Rights::KEY_OLD_ID]) ? $this->rights[$key][Rights::KEY_OLD_ID] : $this->rights[$key][Rights::KEY_ID]);	// permission id in llx_rights_def (not unique because primary key is couple id-entity)
+
+					// Add label with the current language (with specific language key or with PermissionXXX)
+					$labelkey = (isset($this->rights[$key][Rights::KEY_OLD_LABEL]) ? $this->rights[$key][Rights::KEY_OLD_LABEL] : $this->rights[$key][Rights::KEY_LABEL]);
+					$r_label = (getDolGlobalString('MAIN_USE_ADVANCED_PERMS') && ($langs->trans("PermissionAdvanced".$r_id) != "PermissionAdvanced".$r_id) ? $langs->trans("PermissionAdvanced".$r_id) : (($langs->trans("Permission".$r_id) != "Permission".$r_id) ? $langs->trans("Permission".$r_id) : $langs->trans($labelkey)));
+
+					$r_type		= isset($this->rights[$key][2]) ? $this->rights[$key][2] : '';	// TODO deprecated
+
+					$r_default	= (empty($this->rights[$key][Rights::KEY_OLD_DEFAULT]) ? (!isset($this->rights[$key][Rights::KEY_DEFAULT]) ? 0 : ((int) $this->rights[$key][Rights::KEY_DEFAULT])) : ((int) $this->rights[$key][3]));
```
Comment (mdeweerd) [inline htdocs/core/modules/DolibarrModules.class.php:1904]: Why not have a constant for index 3 as well ?

**Comment by hregis** [comment]
@eldy it's ok for you for the first step?

**Comment by hregis** [comment]
@eldy it's ready for the first step... as a reminder, the new "module_origin" field will allow you to deactivate the rights added to a module if you deactivate the original module. And it will also allow you to put a different icon on the rights of a module coming from another module. thanks

**Comment by mdeweerd** [comment]
Maybe better than origin: parent, root, dependenc(y/ies), ancestor, seed_module(s) .

**Comment by hregis** [comment]
@mdeweerd yes it's like it would be better if the earth rotated less quickly to live longer? ;-)

**Comment by mdeweerd** [comment]
> @mdeweerd yes it's like it would be better if the earth rotated less quickly to live longer? ;-)

I am suggesting alternatives because of @eldy 's comment and how this can be clear to other developers.

BTW: if the earth rotates slower, we do not live more earth days if that does not impact our lifetime in number of oscillations of the caesium atom (which is used to define how long 1 second is).

**Comment by hregis** [comment]
@mdeweerd @eldy I can rename "module_origin" to "parent" if you think it's more meaningful?

**Pull request merged**